pytorch_lightning/loops/epoch_loop.py

```diff
@@ -0,0 +1,255 @@
+import logging
+from contextlib import suppress
+from copy import deepcopy
+from typing import Any, List, Optional, Tuple
+
+import torch
+from torch.optim import Optimizer
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks.early_stopping import EarlyStopping
+from pytorch_lightning.loops.base import Loop
+from pytorch_lightning.loops.training_loop import TrainingLoop
+from pytorch_lightning.trainer.connectors.logger_connector.result import ResultCollection
+from pytorch_lightning.trainer.supporters import TensorRunningAccum
+from pytorch_lightning.utilities import rank_zero_info
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.model_helpers import is_overridden
+from pytorch_lightning.utilities.parsing import AttributeDict
+from pytorch_lightning.utilities.warnings import WarningCache
+
+log = logging.getLogger(__name__)
+
+
+# TODO: typing
+class EpochLoop(Loop):
+
+    def __init__(self, min_epochs, max_epochs, min_steps, max_steps):
+        super().__init__()
+        self._teardown_already_run = False
+
+        # TODO: Move this to trainer (it's a trainer default, loops shouldn't have to care about this
+        # If neither max_epochs or max_steps is set, then use existing default of max_epochs = 1000
+        self.max_epochs = 1000 if (max_epochs is None and max_steps is None) else max_epochs
+        # If neither min_epochs or min_steps is set, then use existing default of min_epochs = 1
+        self.min_epochs = 1 if (min_epochs is None and min_steps is None) else min_epochs
+
+        self.training_loop = TrainingLoop(min_steps, max_steps)
+
+        self.train_results = ResultCollection(True)
+
+    @property
+    def current_epoch(self) -> int:
+        return self.iteration_count
+
+    @current_epoch.setter
+    def current_epoch(self, value: int):
+        self.iteration_count = value
+
+    @property
+    def global_step(self):
+        return self.training_loop.global_step
+
+    @global_step.setter
+    def global_step(self, value):
+        self.training_loop.global_step = value
+
+    @property
+    def total_batch_idx(self):
+        return self.training_loop.total_batch_idx
+
+    @property
+    def batch_idx(self):
+        return self.training_loop.iteration_count
+
+    @property
+    def split_idx(self):
+        return self.training_loop.split_idx
+
+    @property
+    def min_steps(self):
+        return self.training_loop.min_steps
+
+    @property
+    def max_steps(self):
+        return self.training_loop.max_steps
+
+    @max_steps.setter
+    def max_steps(self, value):
+        # TODO: This setter is required by debugging connector (fast dev run)
+        self.training_loop.max_steps = value
```

Comment:
could fast_dev_run be an input arg in the constructor?

Reply:
We could do that, but that just means we add logic that is not necessarily related to the loop in here. The loop has a number of max iterations to run. Where this comes from shouldn't matter for the loop. IMO that's exactly what connectors are for